aya-log-parser/src/lib.rs

```diff
@@ -0,0 +1,182 @@
+fn parse_display_hint(s: &str) -> Result<DisplayHint, String> {
+    Ok(match s {
+        "x" => DisplayHint::LowerHex,
+        "X" => DisplayHint::UpperHex,
+        "ipv4" => DisplayHint::IPv4,
+        "IPv4" => DisplayHint::IPv4,
+        "ipv6" => DisplayHint::IPv6,
+        "IPv6" => DisplayHint::IPv6,
+        _ => return Err(format!("unknown display hint: {:?}", s)),
+    })
+}
+
+/// Parse `Param` from the given `&str` which can specify an optional format
+/// like `:x` or `:ipv4` (without curly braces, which are parsed by the `parse`
+/// function).
+fn parse_param(mut input: &str) -> Result<Parameter, String> {
+    const HINT_PREFIX: &str = ":";
+
+    // Then, optional hint
+    let mut hint = DisplayHint::Default;
+
+    if input.starts_with(HINT_PREFIX) {
+        // skip the prefix
+        input = &input[HINT_PREFIX.len()..];
+        if input.is_empty() {
+            return Err("malformed format string (missing display hint after ':')".into());
+        }
+
+        hint = parse_display_hint(input)?;
+    } else if !input.is_empty() {
+        return Err(format!("unexpected content {:?} in format string", input));
+    }
+
+    Ok(Parameter { hint })
+}
+
+/// Parses the given format string into string literals and parameters specified
+/// by curly braces (with optional format hints like `:x` or `:ipv4`).
+pub fn parse(format_string: &str) -> Result<Vec<Fragment>, String> {
+    let mut fragments = Vec::new();
+
+    // Index after the `}` of the last format specifier.
+    let mut end_pos = 0;
+
+    let mut chars = format_string.char_indices();
+    while let Some((brace_pos, ch)) = chars.next() {
+        if ch != '{' {
+            // Part of a literal fragment.
+            continue;
+        }
+
+        // Peek at the next char.
+        if chars.as_str().starts_with('{') {
+            // Escaped `{{`, also part of a literal fragment.
+            chars.next();
+            continue;
+        }
+
+        if brace_pos > end_pos {
+            // There's a literal fragment with at least 1 character before this
+            // parameter fragment.
+            let unescaped_literal = &format_string[end_pos..brace_pos];
+            push_literal(&mut fragments, unescaped_literal)?;
+        }
+
+        // Else, this is a format specifier. It ends at the next `}`.
+        let len = chars
+            .as_str()
+            .find('}')
+            .ok_or("missing `}` in format string")?;
+        end_pos = brace_pos + 1 + len + 1;
+
+        // Parse the contents inside the braces.
+        let param_str = &format_string[brace_pos + 1..][..len];
+        let param = parse_param(param_str)?;
+        fragments.push(Fragment::Parameter(param));
+    }
+
+    // Trailing literal.
+    if end_pos != format_string.len() {
+        push_literal(&mut fragments, &format_string[end_pos..])?;
+    }
+
+    Ok(fragments)
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+
+    #[test]
+    fn test_parse() {
```

Comment:
I think we should have some tests for malformed inputs?

Reply:
I added more tests.